ganttproject/src/net/sourceforge/ganttproject/action/resource/AssignmentToggleAction.java

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2019 BarD Software s.r.o, Juanan Pereira
+
+This file is part of GanttProject, an opensource project management tool.
+
+GanttProject is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+GanttProject is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
+*/
+package net.sourceforge.ganttproject.action.resource;
+
+import net.sourceforge.ganttproject.action.GPAction;
+import net.sourceforge.ganttproject.gui.UIFacade;
+import net.sourceforge.ganttproject.resource.HumanResource;
+import net.sourceforge.ganttproject.task.ResourceAssignment;
+import net.sourceforge.ganttproject.task.ResourceAssignmentMutator;
+import net.sourceforge.ganttproject.task.Task;
+
+import javax.swing.*;
+import java.awt.event.ActionEvent;
+
+
+/**
+ * Action that adds/removes the assignment of a task to a resource
+ */
+public class AssignmentToggleAction extends GPAction {
+  private final HumanResource myHumanResource;
+  private final Task myTask;
+  private final UIFacade myUIFacade;
+
+  public AssignmentToggleAction(HumanResource hr, Task task, UIFacade uiFacade) {
+    super(hr.getName(), IconSize.TOOLBAR_SMALL);
+    myHumanResource = hr;
+    myTask = task;
+    myUIFacade = uiFacade;
+  }
+
+  private void createAssignment() {
+    ResourceAssignmentMutator mutator = myTask.getAssignmentCollection().createMutator();
+    ResourceAssignment newAssignment = mutator.addAssignment(myHumanResource);
+    newAssignment.setLoad(100);
+    newAssignment.setRoleForAssignment(newAssignment.getResource().getRole());
+    mutator.commit();
+  }
+
+  public void delete(HumanResource hr) {
+    ResourceAssignmentMutator mutator = myTask.getAssignmentCollection().createMutator();
+    mutator.deleteAssignment(hr);
+    mutator.commit();
+  }
+
+  @Override
+  public void actionPerformed(ActionEvent e) {
+    Boolean isSelected = (Boolean) getValue(Action.SELECTED_KEY);
+    myUIFacade.getUndoManager().undoableEdit(getLocalizedDescription(), new Runnable() {
+      @Override
+      public void run() {
+        if (isSelected == null) {
```

Comment:
According to IntelliJ, a casting to Boolean is needed for this to work. Otherwise I get this compilation error:

> Operator '!=' cannot be applied to 'java.lang.Object', 'boolean'

But if I add a casting to Boolean:

`if ((Boolean)getValue(Action.SELECTED_KEY) != true) {`

then, when _getValue_ returns _null_, I get an npe:

> java.lang.NullPointerException: cannot unbox null value

If the aim is to try to avoid defining the variable from line 63, I can use a line like:

`if (getValue(Action.SELECTED_KEY)==null)`

But in my opinion, the _isSelected_ variable makes the code more readable. Yep, I know, other programmers could argue just the opposite :) What do you think?

Reply:
I just care that no matter what you `put`, you compare with `null`. At the moment you `put` Boolean.TRUE which makes menu item selected, however, you might as well `put` `Boolean.FALSE` (which means the action is NOT selected) which would not have any effect here: it is still != null.
Replacing atomic `true` with `Boolean.TRUE` should solve the issue with casting and NPEs.

Reply:
Thanks for the explanation. I think that I get it. I have patched two things in my last commit:

* Initialize the AssignmentToggleAction `SELECTED_KEY` to either true or false, accordingly, to avoid comparing with null. Previously, I was only initializing that key to true.
* Replace the lines of code related to this comment with:

`if (getValue(Action.SELECTED_KEY) == Boolean.TRUE) {
  createAssignment();
} else {
  delete(myHumanResource);
}`

This is weird (for me), because I expected to get a Boolean.FALSE when a resource is not selected (and I clicked to, well, select it!) but -using the debugger- it seems that when I click over the name of a human resource to assign it to a certain task,  Swing is changing immediately the value of the SELECTED_KEY to Boolean.TRUE, thus it is what I obtain when I ask for it inside the actionPerformed method.